mongodbatlas/fw_resource_mongodbatlas_search_deployment_state_transition_test.go

```diff
@@ -0,0 +1,204 @@
+package mongodbatlas
+
+import (
+	"context"
+	"errors"
+	"log"
+	"net/http"
+	"reflect"
+	"testing"
+	"time"
+
+	"github.com/mongodb/terraform-provider-mongodbatlas/mongodbatlas/framework/retry"
+	"go.mongodb.org/atlas-sdk/v20231115001/admin"
+)
+
+type stateTransitionTestCase struct {
+	expectedResult *admin.ApiSearchDeploymentResponse
+	name           string
+	mockResponses  []SearchDeploymentResponse
+	expectedError  bool
+}
+
+func TestSearchDeplomentStateTransition(t *testing.T) {
+	testCases := []stateTransitionTestCase{
+		{
+			name: "Successful transition to IDLE",
+			mockResponses: []SearchDeploymentResponse{
+				{
+					DeploymentResp: responseWithState("UPDATING"),
+				},
+				{
+					DeploymentResp: responseWithState("IDLE"),
+				},
+			},
+			expectedResult: responseWithState("IDLE"),
+			expectedError:  false,
+		},
+		{
+			name: "Successful transition to IDLE with 503 error in between",
+			mockResponses: []SearchDeploymentResponse{
+				{
+					DeploymentResp: responseWithState("UPDATING"),
+				},
+				{
+					DeploymentResp: nil,
+					HTTPResponse:   &http.Response{StatusCode: 503},
+					Err:            errors.New("Service Unavailable"),
+				},
+				{
+					DeploymentResp: responseWithState("IDLE"),
+				},
+			},
+			expectedResult: responseWithState("IDLE"),
+			expectedError:  false,
+		},
+		{
+			name: "Error when transitioning to an unknown state",
+			mockResponses: []SearchDeploymentResponse{
+				{
+					DeploymentResp: responseWithState("UPDATING"),
+				},
+				{
+					DeploymentResp: responseWithState(""),
+				},
+			},
+			expectedResult: nil,
+			expectedError:  true,
+		},
+		{
+			name: "Error when API responds with error",
+			mockResponses: []SearchDeploymentResponse{
+				{
+					DeploymentResp: nil,
+					HTTPResponse:   &http.Response{StatusCode: 500},
+					Err:            errors.New("Internal server error"),
+				},
+			},
+			expectedResult: nil,
+			expectedError:  true,
+		},
+	}
+
+	for _, tc := range testCases {
+		t.Run(tc.name, func(t *testing.T) {
+			mockService := MockSearchDeploymentService{
+				MockResponses: tc.mockResponses,
+			}
+
+			resp, err := waitSearchNodeStateTransition(context.Background(), dummyID, "Cluster0", &mockService, testTimeoutConfig)
+
+			if (err != nil) != tc.expectedError {
+				t.Errorf("Case %s: Received unexpected error: %v", tc.name, err)
+			}
+
+			if !reflect.DeepEqual(tc.expectedResult, resp) {
+				t.Errorf("Case %s: Response did not match expected output", tc.name)
+			}
+		})
+	}
+}
+
+type deleteStateTransitionTestCase struct {
```

Comment:
Can't you reuse `stateTransitionTestCase`? in other work, do you need two structs or only 1?

Reply:
Yep was able to reuse here. Simply do not use/define the field `expectedResult *admin.ApiSearchDeploymentResponse` for delete tests. Thanks!